### PR TITLE
feat: [SMP-2375]: Added helper function to create init containers

### PR DESCRIPTION
**Pull request by avaakash**
This helper function will add the wait-for initContainers based on
1. If waitForInitContainers in values is provided (local reference) it will be preferred
2. If not, then global.waitForInitContainer will be preferred

### CODE REVIEW

File: src/common/templates/_initContainer.tpl [hunk 1]
```diff
@@ -18,5 +18,63 @@ Params:
...
+Create initContainer with wait for label app
+Usage: {{ include "common.initContainer.waitForContainer" (dict "root" . "containerName" "container-name" "appName" "app-name") }}
+
+Params:
+  - values - Object - Required. helm values
+  - containerName - String - Required. name of the container to set
+  - appName - String - Required. name of the app to wait for
+*/}}
+
+{{- define "migrator.initContainer.waitForContainer" -}}
```
Comment (arya-harness) [inline src/common/templates/_initContainer.tpl:36]: fix function name here

**Review by arya-harness — APPROVED** [review]

**Comment by arya-harness** [comment]
How can we solve for cases where we want to disable only one initcontainer without affecting other charts?

**Comment by avaakash** [comment]
Right now, all wait-for init containers for one chart can be disabled.
If we want to disable one specific container, say we have to wait-for containers (wait-for-mongo, wait-for-redis) in one service (say ng-manager) and we only want to disable wait-for-mongo, achieving this will require creating a new setting for each wait-for container. This is doable but will add more complexity.
We can based on it, do this in the chart
```
{{- if .Values.waitForInitContainer.mongo }}
--- include container ---
{{- end }}
```
And in values.yaml
```
waitForInitContainer:
    mongo: true
    redis: false
```
Something like this can be done

File: src/common/templates/_initContainer.tpl [hunk 2]
```diff
@@ -18,5 +18,63 @@ Params:
...
+
+{{- define "common.initContainer.waitForContainer" -}}
+{{- $values := .root.Values }}
+{{- $local := $values.waitForInitContainer }}
+{{- $global := $values.global.waitForInitContainer }}
+{{- $containerNameDerived := printf "wait-for-%s" $.appName }}
+{{- if .containerName }}
+  {{- $containerNameDerived = $.containerName }}
+{{- end }}
+
```
Comment (kapilgarg1996) [inline src/common/templates/_initContainer.tpl:44]: remove this empty line

File: src/common/templates/_initContainer.tpl [hunk 3]
```diff
@@ -18,5 +18,63 @@ Params:
...
+  {{- end }}
+  {{- if $local.containerSecurityContext }}
+  securityContext:
+    {{- toYaml $local.containerSecurityContext | nindent 4 }}
+  {{- end }}
+  args:
+    - "pod"
+    - "-lapp={{ .appName }}"
+  {{- end }}
+{{- else }}
```
Comment (kapilgarg1996) [inline src/common/templates/_initContainer.tpl:62]: shouldn't we merge local and global waitFor containers ?

**Comment by avaakash** [comment]
@kapilgarg1996 can you review again, have made the changes. I have used mergeOverwrite, tested it with multiple scenarios between global/local values.

File: src/common/templates/_initContainer.tpl [hunk 4]
```diff
@@ -18,5 +18,43 @@ Params:
...
+  - appName - String - Required. name of the app to wait for
+*/}}
+
+{{- define "common.initContainer.waitForContainer" -}}
+{{- $values := .root.Values }}
+{{- $local := $values.waitForInitContainer }}
+{{- $global := $values.global.waitForInitContainer }}
+{{- $containerNameDerived := printf "wait-for-%s" $.appName }}
+{{- if .containerName }}
+  {{- $containerNameDerived = $.containerName }}
```
Comment (kapilgarg1996) [inline src/common/templates/_initContainer.tpl:41]: this seems wrong. should be .containerName

**Review by kapilgarg1996 — APPROVED** [review]

**Pull request merged**